File: sorting/15-3Sum.py
# 15. 3Sum
# 17 March 2023
# https://leetcode.com/problems/3sum/
# David Pratt

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        map = {}
        for i in range(n):
            map[nums[i]] = i

        found = set()
        result = []

        for i in range(0, n-1):
            for j in range(i+1, n):
                x = nums[i]
                y = nums[j]
                z = -1 * (x+y) # x+y = -z

                if z in map and i != map[z] and j != map[z]:
                    trip = [x, y, z]
                    trip.sort()
                    if str(trip) in found:
                        continue
                    else:
                        found.add(str(trip))
                        result.append(trip)
        return result

File: array/15-3-Sum.js
/*
15. 3 Sum
https://leetcode.com/problems/3sum/
5 April 2023
David Pratt
*/

var threeSum = function (nums) {
  let result = [];
  nums.sort((a, b) => a - b);
  for (let i = 0; i < nums.length - 2; i++) {
    if (nums[i] > 0) break;
    if (i > 0 && nums[i] == nums[i - 1]) continue;
    let left = i + 1;
    let right = nums.length - 1;
    while (left < right) {
      const sum = nums[i] + nums[left] + nums[right];
      if (sum < 0) left++;
      else if (sum > 0) right--;
      else {
        result.push([nums[i], nums[left], nums[right]]);
        right--;
        left++;
        while (left < right && nums[left] == nums[left - 1]) left++;
        while (left < right && nums[right] == nums[right + 1]) right--;
      }
    }
  }
  return result;
};
